File: market/test/integration/java/ru/yandex/market/logistics/utilizer/security/tvm/TvmSecurityTest.java
package ru.yandex.market.logistics.utilizer.security.tvm;

import com.github.springtestdbunit.annotation.DatabaseSetup;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.context.annotation.Import;
import org.springframework.test.context.TestPropertySource;

import ru.yandex.market.logistics.util.client.tvm.client.TvmClientApi;
import ru.yandex.market.logistics.utilizer.base.AbstractContextualTest;
import ru.yandex.market.logistics.utilizer.config.SecurityTestConfig;

import static org.hamcrest.CoreMatchers.containsString;
import static org.hamcrest.Matchers.isEmptyOrNullString;
import static org.hamcrest.Matchers.not;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
import static ru.yandex.market.logistics.utilizer.security.tvm.TvmTestUtils.baseRequestBuilder;
import static ru.yandex.market.logistics.utilizer.security.tvm.TvmTestUtils.builderWithEmptyServiceTicket;
import static ru.yandex.market.logistics.utilizer.security.tvm.TvmTestUtils.builderWithServiceTicket;
import static ru.yandex.market.logistics.utilizer.security.tvm.TvmTestUtils.builderWithUserTicket;
import static ru.yandex.market.logistics.utilizer.security.tvm.TvmTestUtils.mockServiceTicketCheck;
import static ru.yandex.market.logistics.utilizer.security.tvm.TvmTestUtils.mockUserTicketCheck;


@DatabaseSetup("classpath:fixtures/controller/utilization-cycle/warehouse-stock-types/3/before.xml")
@TestPropertySource(properties = {
        "tvm.unsecured-methods=/actuator/health",
        "tvm.utilizer.check-user-ticket=true",
})
public class TvmSecurityTest extends AbstractContextualTest {

    @Autowired
    private TvmClientApi tvmClientApi;

    @Test
    public void testRequestIsAbortedWithoutTicket() throws Exception {
        mockMvc.perform(baseRequestBuilder())
                .andExpect(status().is4xxClientError())
                .andExpect(content().string(containsString("SERVICE_TICKET_NOT_PRESENT")));
    }


    @Test
    public void testRequestIsSuccessfulForUnsecuredActuatorHealth() throws Exception {
        mockMvc.perform(get("/actuator/health"))
                .andExpect(status().isOk())
                .andExpect(content().json("{\"status\": \"UP\"}"));
    }

    @Test
    public void testRequestIsAbortedWithEmptyServiceTicket() throws Exception {
        mockMvc.perform(builderWithEmptyServiceTicket())
                .andExpect(status().is4xxClientError())
                .andExpect(content().string(containsString("SERVICE_TICKET_NOT_PRESENT")));
    }

    @Test
    public void testRequestIsAbortedWithNotValidServiceTicket() throws Exception {

        mockServiceTicketCheck(false, tvmClientApi);

        mockMvc.perform(builderWithServiceTicket())
                .andExpect(status().is4xxClientError())
                .andExpect(content().string(containsString("BAD_SERVICE_TICKET")));
    }

    @Test
    public void testSuccessfulRequestWithUserTicket() throws Exception {

        mockServiceTicketCheck(true, tvmClientApi);
        mockUserTicketCheck(true, tvmClientApi);

        mockMvc.perform(builderWithUserTicket())
                .andExpect(status().is2xxSuccessful())
                .andExpect(content().string(not(isEmptyOrNullString())));
    }

    @Test
    public void testRequestIsAbortedWithNotValidUserTicket() throws Exception {

        mockServiceTicketCheck(true, tvmClientApi);
        mockUserTicketCheck(false, tvmClientApi);

        mockMvc.perform(builderWithUserTicket())
                .andExpect(status().is4xxClientError())
                .andExpect(content().string(containsString("BAD_USER_TICKET")));

    }
}

File: market/test/integration/kotlin/tvm/TvmSecurityTest.kt
package ru.yandex.market.logistics.calendaring.tvm

import com.github.springtestdbunit.annotation.DatabaseSetup
import org.hamcrest.CoreMatchers
import org.hamcrest.Matchers
import org.hamcrest.Matchers.emptyOrNullString
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.Import
import org.springframework.test.context.TestPropertySource
import org.springframework.test.web.servlet.result.MockMvcResultMatchers
import ru.yandex.market.logistics.calendaring.base.AbstractContextualTest
import ru.yandex.market.logistics.calendaring.config.SecurityTestConfig
import ru.yandex.market.logistics.calendaring.tvm.TvmTestUtils.Companion.baseRequestBuilder
import ru.yandex.market.logistics.calendaring.tvm.TvmTestUtils.Companion.builderWithEmptyServiceTicket
import ru.yandex.market.logistics.calendaring.tvm.TvmTestUtils.Companion.builderWithServiceTicket
import ru.yandex.market.logistics.calendaring.tvm.TvmTestUtils.Companion.builderWithUserTicket
import ru.yandex.market.logistics.calendaring.tvm.TvmTestUtils.Companion.mockServiceTicketCheck
import ru.yandex.market.logistics.calendaring.tvm.TvmTestUtils.Companion.mockUserTicketCheck
import ru.yandex.market.logistics.util.client.tvm.client.TvmClientApi


@TestPropertySource(properties = ["tvm.unsecured-methods=/ping", "tvm.internal.check-user-ticket=true"])
@Import(SecurityTestConfig::class)
class TvmSecurityTest(
    @Autowired
    private val tvmClientApi: TvmClientApi
) : AbstractContextualTest() {

    @Test
    @Throws(Exception::class)
    fun testRequestIsAbortedWithoutTicket() {
        mockMvc!!.perform(baseRequestBuilder())
            .andExpect(MockMvcResultMatchers.status().is4xxClientError)
            .andExpect(
                MockMvcResultMatchers.content().string(CoreMatchers.containsString("SERVICE_TICKET_NOT_PRESENT"))
            )
    }

    @Test
    @Throws(Exception::class)
    fun testRequestIsAbortedWithEmptyServiceTicket() {
        mockMvc!!.perform(builderWithEmptyServiceTicket())
            .andExpect(MockMvcResultMatchers.status().is4xxClientError)
            .andExpect(
                MockMvcResultMatchers.content().string(CoreMatchers.containsString("SERVICE_TICKET_NOT_PRESENT"))
            )
    }

    @Test
    @Throws(Exception::class)
    fun testRequestIsAbortedWithNotValidServiceTicket() {
        mockServiceTicketCheck(false, tvmClientApi)
        mockMvc!!.perform(builderWithServiceTicket())
            .andExpect(MockMvcResultMatchers.status().is4xxClientError)
            .andExpect(MockMvcResultMatchers.content().string(CoreMatchers.containsString("BAD_SERVICE_TICKET")))
    }

    @Test
    @Throws(Exception::class)
    @DatabaseSetup("classpath:fixtures/controller/booking/get-slot/slot-exists/before.xml")
    fun testSuccessfulRequestWithUserTicket() {
        mockServiceTicketCheck(true, tvmClientApi)
        mockUserTicketCheck(true, tvmClientApi)
        mockMvc!!.perform(builderWithUserTicket())
            .andExpect(MockMvcResultMatchers.status().is2xxSuccessful)
            .andExpect(MockMvcResultMatchers.content().string(Matchers.not(Matchers.`is`(emptyOrNullString()))))
    }

    @Test
    @Throws(Exception::class)
    fun testRequestIsAbortedWithNotValidUserTicket() {
        mockServiceTicketCheck(true, tvmClientApi)
        mockUserTicketCheck(false, tvmClientApi)
        mockMvc!!.perform(builderWithUserTicket())
            .andExpect(MockMvcResultMatchers.status().is4xxClientError)
            .andExpect(MockMvcResultMatchers.content().string(CoreMatchers.containsString("BAD_USER_TICKET")))
    }
}
